### ISSUE TITLE
See how we can further optimize fast sync

### ISSUE THREAD

**Issue by gsalgado** [ethereum/py-evm#1065]
Breakdown of important events during a fast-sync import of a 192-block batch:

```
   DEBUG  07-21 08:43:33       chain  Fetching chain segment starting at #3919932
   DEBUG  07-21 08:43:33       chain  Got BlockHeaders from 3919932 to 3920123
   DEBUG  07-21 08:43:34       chain  Got new header chain starting at #3919932

# Here we have already validated (including PoW verification) the chain segment above
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0xcde6@13.95.123.161)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x38b0@136.179.11.188)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x6914@35.172.216.29)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x4228@54.86.13.47)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x590b@103.75.0.132)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0xb969@47.52.27.7)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x3aaf@47.88.228.205)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0xfede@89.38.99.153)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0x952e@18.188.70.61)>
   DEBUG  07-21 08:43:34       chain  Requesting 19 block bodies to ETHPeer <Node(0xc496@40.84.48.167)>

# Each of these are triggered when we receive a BlockBodies msg, and they will run in parallel,
# generating the tx trie (in the process pool executor) and persisting them to the DB
   DEBUG  07-21 08:43:34       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x4228@54.86.13.47)>
   DEBUG  07-21 08:43:34       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x38b0@136.179.11.188)>
   DEBUG  07-21 08:43:34       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x6914@35.172.216.29)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x952e@18.188.70.61)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0xc496@40.84.48.167)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0xb969@47.52.27.7)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0xcde6@13.95.123.161)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0xfede@89.38.99.153)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x3aaf@47.88.228.205)>
   DEBUG  07-21 08:43:35       chain  Got Bodies for 19 blocks from ETHPeer <Node(0x590b@103.75.0.132)>

# This means it took ~5s for all the coroutines above to complete.
   DEBUG  07-21 08:43:39       chain  Got block bodies for chain segment

# Now we request receipts...
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0xcde6@13.95.123.161)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x38b0@136.179.11.188)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x6914@35.172.216.29)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x4228@54.86.13.47)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x590b@103.75.0.132)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0xb969@47.52.27.7)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x3aaf@47.88.228.205)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0xfede@89.38.99.153)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0x952e@18.188.70.61)>
   DEBUG  07-21 08:43:39       chain  Requesting 19 block receipts to ETHPeer <Node(0xc496@40.84.48.167)>

# And spawn coroutines to process/persist them as well
   DEBUG  07-21 08:43:39       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x4228@54.86.13.47)>
   DEBUG  07-21 08:43:39       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x952e@18.188.70.61)>
   DEBUG  07-21 08:43:40       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x38b0@136.179.11.188)>
   DEBUG  07-21 08:43:40       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x6914@35.172.216.29)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0xb969@47.52.27.7)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0xc496@40.84.48.167)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0xcde6@13.95.123.161)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x590b@103.75.0.132)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0xfede@89.38.99.153)>
   DEBUG  07-21 08:43:41       chain  Got Receipts for 19 blocks from ETHPeer <Node(0x3aaf@47.88.228.205)>

# Here all coroutines above completed: 3s 
   DEBUG  07-21 08:43:44       chain  Got block receipts for chain segment

# Now we combine the downloaded bodies with the headers and persist them to the DB: 
# https://github.com/ethereum/py-evm/blob/4b3e30b4edf0e3ca0b6ddd48dca469dfa8115c52/p2p/chain.py#L378
# This runs on the main thread, but the heavy lifting is done in the DB process
   
# This would mean ~4s on all 192 ChainDB.coro_persist_block() calls
   INFO  07-21 08:43:48       chain  Imported 192 headers in 14.43 seconds, new head: #3920123
```

So, ~1s to download bodies, ~2s to download receipts and ~12 to process/persist them.

**Comment by gsalgado** [comment]
Full log file: [trinity.log](https://github.com/ethereum/py-evm/files/2216039/trinity.log)

**Comment by pipermerriam** [comment]
One of the optimizations that I've been thinking about is sharding our chain data.

Simple implementation is to only store 1/N blocks.  My understanding is the 90% or more of the hard drive footprint is Blocks/Receipts/Transactions/Headers.

So potentially, we can speed up sync by some non-trivial amount by just not fetching all of the block data.

I'll write up an issue to capture this idea more fully.

**Comment by pipermerriam** [comment]
https://github.com/ethereum/py-evm/issues/1066

**Comment by gsalgado** [comment]
Now for some profiling data:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    32076    0.503    0.000  133.752    0.004 base_events.py:1336(_run_once)
    32076    0.167    0.000   90.054    0.003 selectors.py:428(select)
    32076   89.822    0.003   89.822    0.003 {method 'poll' of 'select.epoll' objects}
   100332    0.333    0.000   42.994    0.000 events.py:143(_run)
815791/61951    1.567    0.000   32.137    0.001 functional.py:20(inner)
     7944    0.014    0.000   24.057    0.003 integration_test_helpers.py:28(passthrough_method)
   274973    0.244    0.000   23.687    0.000 codec.py:194(decode)
     1895    0.076    0.000   20.390    0.011 chain.py:263(persist_block)
      320    0.003    0.000   16.709    0.052 peer.py:334(read_msg)
      151    0.000    0.000   16.696    0.111 service.py:45(run)
      141    0.003    0.000   16.693    0.118 peer.py:319(_run)
      106    0.005    0.000   16.357    0.154 protocol.py:93(decode)
      106    0.020    0.000   16.350    0.154 protocol.py:78(decode_payload)
    45860    0.069    0.000   15.803    0.000 chain.py:332(get_block_transaction_hashes)
117849/52568    0.443    0.000   14.747    0.000 serializable.py:254(deserialize)
235830/6362    0.312    0.000   14.499    0.002 lists.py:126(deserialize)
    45860    0.137    0.000   14.084    0.000 chain.py:404(_get_block_transaction_data)
    89825    0.175    0.000   12.998    0.000 hexary.py:66(get)
   268612    0.490    0.000    9.556    0.000 hexary.py:186(get_node)
268612/89825    0.394    0.000    8.556    0.000 hexary.py:74(_get)
   274973    0.233    0.000    8.503    0.000 codec.py:182(consume_item)
178574/89574    0.255    0.000    8.318    0.000 hexary.py:386(_get_branch_node)
831891/289814    0.772    0.000    7.615    0.000 lists.py:78(deserialize)
4365910/274973    4.810    0.000    7.444    0.000 codec.py:153(consume_payload)
   266960    0.310    0.000    7.260    0.000 nodes.py:77(decode_node)
    45860    0.029    0.000    6.936    0.000 hexary.py:423(__contains__)
    45860    0.056    0.000    6.907    0.000 hexary.py:113(exists)
    43965    0.031    0.000    6.204    0.000 hexary.py:414(__getitem__)
    46447    0.327    0.000    5.200    0.000 receipts.py:33(__init__)
   218898    0.976    0.000    4.377    0.000 serializable.py:41(validate_args_and_kwargs)
```

As we can see, most of the time seems to be spent in decoding peer messages and in `ChainDB.persist_block()`, which is to be expected during a fast sync

**Comment by gsalgado** [comment]
The profiling stats above didn't include code executed in subprocesses (e.g. trie generation for transactions/receipts), so it is inaccurate. Here's some stats that include that too:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    459/1    0.008    0.000   72.082   72.082 {built-in method builtins.exec}
        1    0.000    0.000   72.081   72.081 chain.py:770(_test)
        1    0.026    0.026   71.437   71.437 base_events.py:405(run_forever)
    26595    0.380    0.000   71.411    0.003 base_events.py:1336(_run_once)
    97887    0.324    0.000   70.164    0.001 events.py:143(_run)
     6009    0.004    0.000   38.289    0.006 chain.py:115(handle_msg)
     6009    0.005    0.000   38.285    0.006 chain.py:506(_handle_msg)
       37    0.006    0.000   37.704    1.019 chain.py:851(mock_run_in_executor)
     3991    0.055    0.000   37.698    0.009 trie.py:16(make_trie_root_and_nodes)
     3981    0.140    0.000   32.375    0.008 trie.py:25(_make_trie_root_and_nodes)
    43777    0.046    0.000   30.795    0.001 hexary.py:417(__setitem__)
    43777    0.148    0.000   30.750    0.001 hexary.py:86(set)
110954/43777    0.262    0.000   22.825    0.001 hexary.py:96(_set)
643253/220392    1.389    0.000   22.364    0.000 functional.py:20(inner)
     4058    0.019    0.000   20.087    0.005 chain.py:561(_handle_block_bodies)
   193699    0.417    0.000   19.202    0.000 hexary.py:200(_node_to_db_mapping)
     8272    0.017    0.000   19.006    0.002 integration_test_helpers.py:28(passthrough_method)
     1863    0.009    0.000   18.187    0.010 chain.py:545(_handle_block_receipts)
   216220    0.242    0.000   16.993    0.000 codec.py:194(decode)
3391120/349711    4.311    0.000   16.731    0.000 codec.py:72(encode_raw)
67149/36291    0.172    0.000   16.274    0.000 hexary.py:321(_set_branch_node)
320795/277555    1.073    0.000   13.100    0.000 {method 'join' of 'bytes' objects}
     2279    0.058    0.000   12.974    0.006 chain.py:263(persist_block)
3362203/3246243    1.321    0.000   12.080    0.000 codec.py:80(<genexpr>)
10456070/10455385    4.653    0.000   10.683    0.000 {built-in method builtins.isinstance}
   110954    0.185    0.000   10.519    0.000 hexary.py:218(_prune_node)
      272    0.001    0.000   10.056    0.037 service.py:45(run)
      260    0.001    0.000   10.053    0.039 peer.py:319(_run)
      395    0.004    0.000   10.045    0.025 peer.py:334(read_msg)
   268161    0.555    0.000    9.942    0.000 main.py:32(__call__)
      131    0.002    0.000    9.851    0.075 protocol.py:93(decode)
      131    0.013    0.000    9.847    0.075 protocol.py:78(decode_payload)
   205486    0.326    0.000    9.700    0.000 crypto.py:8(keccak)
    82745    0.161    0.000    9.166    0.000 hexary.py:212(_persist_node)
61746/28290    0.277    0.000    9.122    0.000 serializable.py:254(deserialize)
```

**Comment by gsalgado** [comment]
Closing in favour of specific issues: #1102 and #1123
